File: mobile/ios/SpeakeasyCrypto/Vault/VaultKeyManager.swift
import Foundation
import CryptoKit

// Derives Vault Key (VK) from Device Master Key (DMK) via HKDF

class VaultKeyManager {
    
    private let store: SpeakeasyStore
    private let hkdfInfo = Data("speakeasy-vault-v1".utf8)
    private let hkdfSalt = Data("speakeasy".utf8) // Stable salt
    
    init(store: SpeakeasyStore) {
        self.store = store
    }
    
    // DMK should be stored in Keychain as our identity key or derived from it.
    // For Phase 3, we'll use the identity key itself as the DMK (or derive from it).
    func getDeviceMasterKey() -> SymmetricKey? {
        guard let identityData = store.getIdentityKeyPair() else { return nil }
        // Use first 32 bytes of identity key as DMK (or SHA256 of it)
        let hash = SHA256.hash(data: identityData)
        return SymmetricKey(data: Data(hash))
    }
    
    // Derive Vault Key from DMK
    func getVaultKey() -> SymmetricKey? {
        guard let dmk = getDeviceMasterKey() else { return nil }
        
        // HKDF derivation
        let vkData = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: dmk,
            salt: hkdfSalt,
            info: hkdfInfo,
            outputByteCount: 32
        )
        return vkData
    }
    
    // Generate a random FileKey for each file/attachment
    func generateFileKey() -> SymmetricKey {
        return SymmetricKey(size: .bits256)
    }
    
    // Wrap (encrypt) a FileKey with the VaultKey
    func wrapFileKey(_ fileKey: SymmetricKey) throws -> Data {
        guard let vk = getVaultKey() else {
            throw NSError(domain: "VaultKeyManager", code: 1, userInfo: [NSLocalizedDescriptionKey: "No Vault Key"])
        }
        
        let fileKeyData = fileKey.withUnsafeBytes { Data($0) }
        let box = try AES.GCM.seal(fileKeyData, using: vk)
        return box.combined! // nonce + ciphertext + tag
    }
    
    // Unwrap (decrypt) a FileKey
    func unwrapFileKey(wrapped: Data) throws -> SymmetricKey {
        guard let vk = getVaultKey() else {
            throw NSError(domain: "VaultKeyManager", code: 1, userInfo: [NSLocalizedDescriptionKey: "No Vault Key"])
        }
        
        let box = try AES.GCM.SealedBox(combined: wrapped)
        let fileKeyData = try AES.GCM.open(box, using: vk)
        return SymmetricKey(data: fileKeyData)
    }
}

File: mobile/android/crypto/vault/VaultKeyManager.kt
package com.speakeasy.crypto.vault

import android.content.Context
import android.util.Base64
import java.security.MessageDigest
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import com.speakeasy.crypto.SpeakeasyStore

// Derives Vault Key (VK) from Device Master Key (DMK) via HKDF

class VaultKeyManager(private val store: SpeakeasyStore) {
    
    private val hkdfInfo = "speakeasy-vault-v1".toByteArray()
    private val hkdfSalt = "speakeasy".toByteArray()
    
    // DMK from identity key
    fun getDeviceMasterKey(): ByteArray? {
        val identityData = store.getIdentityKeyPair() ?: return null
        val digest = MessageDigest.getInstance("SHA-256")
        return digest.digest(identityData)
    }
    
    // Derive Vault Key via HKDF
    fun getVaultKey(): SecretKey? {
        val dmk = getDeviceMasterKey() ?: return null
        val derived = hkdf(dmk, hkdfSalt, hkdfInfo, 32)
        return SecretKeySpec(derived, "AES")
    }
    
    // Generate random FileKey
    fun generateFileKey(): SecretKey {
        val keyBytes = ByteArray(32)
        java.security.SecureRandom().nextBytes(keyBytes)
        return SecretKeySpec(keyBytes, "AES")
    }
    
    // Wrap FileKey with VaultKey (AES-GCM)
    fun wrapFileKey(fileKey: SecretKey): ByteArray {
        val vk = getVaultKey() ?: throw IllegalStateException("No Vault Key")
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, vk)
        val iv = cipher.iv
        val encrypted = cipher.doFinal(fileKey.encoded)
        return iv + encrypted // 12-byte IV + ciphertext+tag
    }
    
    // Unwrap FileKey
    fun unwrapFileKey(wrapped: ByteArray): SecretKey {
        val vk = getVaultKey() ?: throw IllegalStateException("No Vault Key")
        val iv = wrapped.sliceArray(0 until 12)
        val ciphertext = wrapped.sliceArray(12 until wrapped.size)
        
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, vk, GCMParameterSpec(128, iv))
        val keyBytes = cipher.doFinal(ciphertext)
        return SecretKeySpec(keyBytes, "AES")
    }
    
    // Simple HKDF implementation (extract + expand)
    private fun hkdf(ikm: ByteArray, salt: ByteArray, info: ByteArray, outputLen: Int): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(salt, "HmacSHA256"))
        val prk = mac.doFinal(ikm)
        
        mac.init(SecretKeySpec(prk, "HmacSHA256"))
        val result = ByteArray(outputLen)
        var t = ByteArray(0)
        var offset = 0
        var i = 1
        while (offset < outputLen) {
            mac.update(t)
            mac.update(info)
            mac.update(i.toByte())
            t = mac.doFinal()
            val toCopy = minOf(t.size, outputLen - offset)
            System.arraycopy(t, 0, result, offset, toCopy)
            offset += toCopy
            i++
        }
        return result
    }
}
